File: regexp-repl/regex-repl.cc
/////////////////////////////////////////////////////////////////////////
// (c) Copyright 2021 IBM Corp. Licensed Materials - Property of IBM.
/////////////////////////////////////////////////////////////////////////

// See https://github.com/proxy-wasm/proxy-wasm-cpp-sdk

#include <string>
#include <unordered_map>
#include <regex>

#include "proxy_wasm_intrinsics.h"

// "root context" class handles lifecycle of the extension
class RegexpReplRoot : public RootContext {
public:
  explicit RegexpReplRoot(uint32_t id, std::string_view root_id)
      : RootContext(id, root_id), strId_(std::to_string(id)) {
    // You probably want to warn at a lower level!
    logWarn("RegexpReplRoot(context_id:" + strId_ + ", \"" + std::string(root_id) + "\")");
  }

  // Called once when the VM loads and once when each hook loads and whenever
  // configuration changes. Returns false if the configuration is invalid.
  virtual bool onConfigure(size_t configuration_size) override;

  // Example showing regex pattern/replace
  // Regular expression, possibly with a capture group, e.g. "banana/([0-9]*)";
  const char *getPattern() { return pattern_; } 
  // A replacement, possibly referencing capture groups, e.g. "status/$1";
  const char *getReplaceWith() {return replaceWith_; }

private:
  std::string strId_;
  const char *pattern_ = NULL;
  const char *replaceWith_ = NULL;
};

// "context" class handles lifecycle of an HTTP request/response
class RegexpRepl : public Context {
public:
  explicit RegexpRepl(uint32_t id, RootContext* root) : Context(id, root), strId_(std::to_string(id)) {
    logWarn("context id: " + strId_ + ": RegexpRepl(...)");
  }

  FilterHeadersStatus onRequestHeaders(uint32_t, bool) override;
  RegexpReplRoot* myRoot() { return (RegexpReplRoot *) this->root(); }

private:
  std::string strId_;
};

// This static constructor registers your classes with the VM at startup
static RegisterContextFactory register_ExampleContext(
  CONTEXT_FACTORY(RegexpRepl),
  ROOT_FACTORY(RegexpReplRoot), 
  "regex-repl");

class LogAtStartup {
public:
  explicit LogAtStartup() {
    logWarn("regex-repl.cc will log this when loaded.");
  }
};
static LogAtStartup dummy;

std::string& boolToString(bool b);

// Called per Envoy worker thread when the VM loads.
bool RegexpReplRoot::onConfigure(size_t configuration_size) {
  logWarn("context_id:" + strId_ + " RegexpReplRoot::onConfigure(" + std::to_string(configuration_size) + ")");

  WasmDataPtr configuration_data = getBufferBytes(WasmBufferType::PluginConfiguration, 0, configuration_size);
  std::string configuration = configuration_data->toString();
  logWarn("ignoring configuration ->" + configuration);

  // TODO read the config as JSON.  Parsing JSON isn't built-in.  Plugins can follow 
  // https://github.com/istio-ecosystem/wasm-extensions/tree/master/extensions/basic_auth
  // as an example.
  pattern_ = "banana/([0-9]*)";
  replaceWith_ = "status/$1";

  return true;
}

// ==================

// Do the actual path rewrite here
FilterHeadersStatus RegexpRepl::onRequestHeaders(uint32_t header_count, bool end_of_stream) {
  logWarn("context_id:" + strId_ + " RegexpRepl::onRequestHeaders(" + std::to_string(header_count) + ", " + boolToString(end_of_stream) + ")");

   // Envoy's "headers" include HTTP headers **and** colon-prefixed Envoy HTTP properties.
   // For example:
   // :authority: httpbin.default:8000
   // :path: /status/200
   // :method: GET
   // user-agent: curl/7.58.0
   // accept: */*
   // x-forwarded-proto: http
   // x-request-id: 1be9883a-3ac3-4dba-9991-1a1301f51ace

  // TODO modify the request headers using the configuration
  char const *key = ":path";
  const char *pattern = myRoot()->getPattern(); // e.g. "banana/([0-9]*)";
  const char *replaceWith = myRoot()->getReplaceWith(); // e.g. "status/$1";

  WasmDataPtr wdpGhmv = getRequestHeader(key);
  std::string sValue = wdpGhmv->toString();

  std::regex re(pattern);
  std::string result = std::regex_replace(sValue, re, replaceWith);
  WasmResult wrRrh = replaceRequestHeader(key, result);
  if (wrRrh == WasmResult::Ok) {
    logWarn("context_id:" + strId_ + " RegexpRepl::onRequestHeaders() replaced " + sValue + " with " + result);
  } else {
    logWarn("context_id:" + strId_ + " RegexpRepl::onRequestHeaders() replaceRequestHeader() FAILED.  wrRrh is " + toString(wrRrh));
  }

  return FilterHeadersStatus::Continue;
}

// ==================

std::string& boolToString(bool b)
{
  static std::string t = "true";
  static std::string f = "false";

  return b ? t : f;
}

File: regexp-repl/regex-repl.ts
/////////////////////////////////////////////////////////////////////////
// (c) Copyright 2021 IBM Corp. Licensed Materials - Property of IBM.
/////////////////////////////////////////////////////////////////////////


// See https://github.com/solo-io/proxy-runtime

export * from "@solo-io/proxy-runtime/proxy"; // this exports the required functions for the proxy to interact with us.
import { RootContext, Context, registerRootContext, 
  FilterHeadersStatusValues, 
  stream_context, 
  LogLevelValues, log } from "@solo-io/proxy-runtime";

// "root context" class handles lifecycle of the extension
class RegexpReplRoot extends RootContext {

  createContext(context_id: u32): Context {
    log(LogLevelValues.warn, "RegexpReplRoot.createContext(context_id:" + this.context_id.toString() + ")");
    log(LogLevelValues.warn, "RegexpReplRoot.createContext() called with pattern set to " + this.getPattern());
    return new RegexpRepl(context_id, this);
  }

  // You may not need an onConfigure() override, as the AssemblyScript SDK keeps the config in memory.

  // (just for logging)
  onDelete(): void {
    log(LogLevelValues.warn, "context id: " + this.context_id.toString() + ": RegexpReplRoot.onDelete()");
    super.onDelete();
  }

  // Example showing regex pattern/replace
  // Regular expression, possibly with a capture group
  getPattern(): string { return "banana/([0-9]*)"; } 
  // A replacement, possibly referencing capture groups
  getReplaceWith(): string {return "status/$1"; }
}

// "context" class handles lifecycle of an HTTP request/response
class RegexpRepl extends Context {

  constructor(context_id: u32, root_context: RegexpReplRoot){
    super(context_id, root_context);

    log(LogLevelValues.warn, "context id: " + this.context_id.toString() + ": RegexpRepl constructor(" + context_id.toString() + ", " + "..." + ")");
  }

  myRoot(): RegexpReplRoot { return (this.root_context as RegexpReplRoot); }

  // onRequestHeaders sees the regular headers plus :authority, :path, and :method
  onRequestHeaders(header_count: u32, end_of_stream: bool): FilterHeadersStatusValues {
    log(LogLevelValues.warn, "context id: " + this.context_id.toString() + ": RegexpRepl.onRequestHeaders(" + header_count.toString() + ", " + end_of_stream.toString() + ")");

    let key = ":path";
    let sValue = stream_context.headers.request.get(key);

    let pattern = this.myRoot().getPattern();
    let replaceWith = this.myRoot().getReplaceWith();
  
    // The next lines don't work, because AssemblyScript doesn't yet support RegExp
    // let re = new RegExp(pattern);
    // let result = sValue.replace(re, replaceWith);
    // So instead we do something simpler for the demo.
    let result = sValue.replace("/banana/", "/status/")
    stream_context.headers.request.replace(key, result);
    // Unlike the C++ API, the AssemblyScript API does not provide a result code

    return super.onRequestHeaders(header_count, end_of_stream)
  }

  onDone(): bool { 
    log(LogLevelValues.warn, "context id: " + this.context_id.toString() + ": RegexpRepl.onDone()");
    return super.onDone();
  }

  onDelete(): void { 
    log(LogLevelValues.warn, "context id: " + this.context_id.toString() + ": RegexpRepl.onDelete()");
    super.onDelete();
  }
}

registerRootContext((context_id: u32) => { return new RegexpReplRoot(context_id); }, "regex-repl");
